### PR TITLE
FQDN L3 aware DNS proxy

### PR DESCRIPTION
**Pull request by raybejjani**
The DNS proxy now accounts for the target L3 selector, only allowing
requests whilelisted for that L3 selector. L4 dependencies are still
left to BPF redirection.

*Backporters*
I have this backported in https://github.com/cilium/cilium/pull/9616.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9587)

### CODE REVIEW

File: pkg/fqdn/dnsproxy/proxy.go [hunk 1]
```diff
@@ -390,6 +421,20 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	}
 }
 
+func lookupDestIPID(destIPStr string) (SecID ipcache.Identity, exists bool, err error) {
+	maxPrefixLen := 32
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: nit:
```suggestion
	maxPrefixLen := 8 * net.IPv4len
```

File: pkg/fqdn/dnsproxy/proxy.go [hunk 2]
```diff
@@ -390,6 +421,20 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
...
+	if ip := net.ParseIP(destIPStr); ip.To4() == nil {
+		maxPrefixLen = 128
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: nit:
```suggestion
		maxPrefixLen = 8 * net.IPv6len
```

File: pkg/fqdn/dnsproxy/proxy.go [hunk 3]
```diff
@@ -390,6 +421,20 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
...
+	}
+
+	for i := maxPrefixLen; exists != true && i >= 0; i-- {
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: We actually store the currently-used prefixes in the daemon somewhere (`prefixCounter`) if you want to go that way; it's probably simpler than importing a full LPM into userspace, particularly for a backport.

File: pkg/proxy/dns.go [hunk 4]
```diff
@@ -106,17 +105,7 @@ func (dr *dnsRedirect) UpdateRules(wg *completion.WaitGroup) (revert.RevertFunc,
...
-				dnsName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-				dnsNameAsRE := matchpattern.ToRegexp(dnsName)
-				DefaultDNSProxy.RemoveAllowed(dnsNameAsRE, fmt.Sprintf("%d", dr.redirect.endpointID))
-
-				dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-				DefaultDNSProxy.RemoveAllowed(dnsPatternAsRE, fmt.Sprintf("%d", dr.redirect.endpointID))
-			}
-		}
+		DefaultDNSProxy.UpdateAllowed(dr.redirect.endpointID, nil)
```
Comment (joestringer) [inline pkg/proxy/dns.go]: This could be a separate commit? Is there a meaningful functional change here?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 5]
```diff
@@ -223,48 +231,52 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
...
-	}
+	// TODO: Add regex precompiles
+	p.allowedByID[endpointID] = newRules
 }
 
 // CheckAllowed checks name against the rules added to the proxy, and only
 // returns true if this endpointID was added (via AddAllowed) previously.
-func (p *DNSProxy) CheckAllowed(name, endpointID string) bool {
-	name = strings.ToLower(name)
+func (p *DNSProxy) CheckAllowed(endpointID uint64, destID identity.NumericIdentity, name string) bool {
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: Is this sufficiently abstracted to unit-test it? Seems like there are unit tests for some cases already, might be able to expand them to test the L7 behaviour?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 6]
```diff
@@ -223,48 +231,52 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
...
+func (p *DNSProxy) UpdateAllowed(endpointID uint64, newRules policy.L7DataMap) {
 	p.Lock()
 	defer p.Unlock()
-	for _, reStr := range reStrToRemove {
-		p.allowed.Remove(reStr, endpointID)
-	}
-	for _, reStr := range reStrToAdd {
-		p.allowed.Add(reStr, endpointID)
-	}
+	// TODO: Add regex precompiles
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: +1

**Review by joestringer — CHANGES_REQUESTED** [review]
For a bugfix, I'm curious about whether you consider a full LPM to be a requirement or whether it's reasonable to get away with only iterating the set of prefixes currently in use (would handle most use cases like only FQDN policies, no CIDR policies).

A few other minor comments/suggestions below. In particular it would be good to see some form of reproducer in unit-test form.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 1 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: oops, yes. I saw these and simply forgot :/

File: pkg/proxy/dns.go (hunk 4 above)
Comment (raybejjani) [reply pkg/proxy/dns.go]: It's to match the change in UpdateAllowed's signature. I could make it a separate commit but then the original wouldn't compile. There shouldn't be a functional change, however, as you suggest.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 5 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: Yes, very much so. In fact, it might be simple to unit test the various rule combos now, instead of doing it in the e2e tests. I'll expand on the tests we have.

File: pkg/proxy/dns.go (hunk 4 above)
Comment (joestringer) [reply pkg/proxy/dns.go]: OK, no point making a commit that won't compile.

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go [hunk 7]
```diff
@@ -311,23 +377,42 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
...
+	}
+	scopedLog.WithField("server", targetServerAddr).Debugf("Found target server to of DNS request secID %+v", serverSecID)
 
 	// The allowed check is first because we don't want to use DNS responses that
 	// endpoints are not allowed to see.
 	// Note: The cache doesn't know about the source of the DNS data (yet) and so
 	// it won't enforce any separation between results from different endpoints.
 	// This isn't ideal but we are trusting the DNS responses anyway.
-	if !p.CheckAllowed(qname, ep.StringID()) {
+	allowed, err := p.CheckAllowed(uint64(ep.ID), targetServerPort, serverSecID.ID, qname)
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go:406]: nit: Any particular reason to cast up the size of epID to uint64 here and below?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 8]
```diff
@@ -95,31 +106,95 @@ type DNSProxy struct {
...
+				dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
+				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
+				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			}
+		}
+		re, err := regexp.Compile(strings.Join(reStrings, "|"))
+		if err != nil {
+			return err
+		}
+		newRE[selector] = re
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go:169]: Interesting, so similar to the `L7RulesPerEp` part of an `L4Filter`, the selector cache reference is reused from the `L4Filter.CachedSelectors`? I suppose that that still makes sense. :+1:

File: pkg/fqdn/dnsproxy/proxy.go [hunk 9]
```diff
@@ -311,23 +377,42 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
...
 	}
-	scopedLog.WithField("server", targetServerAddr).Debug("Found target server to of DNS request")
+
+	serverSecID, exists, err := p.LookupSecIDByIP(targetServerIP)
+	if !exists || err != nil {
+		scopedLog.WithError(err).WithField("server", targetServerAddr).Debug("cannot find server ip in ipcache")
+		stat.Err = fmt.Errorf("Cannot find server ip in ipcache: %s", err)
+		stat.ProcessingTime.End(false)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.sendRefused(scopedLog, w, request)
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go:396]: The ordering of the above few statements for this error handling case is different from the equivalent cases just a few lines down. Is there any particular reason? (Mostly just concerned about consistency of timestamps and so on)

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 10]
```diff
@@ -163,131 +206,325 @@ func (s *DNSProxyTestSuite) TearDownSuite(c *C) {
...
+	//	| All  | DstID1 |      54 | example.com    |
+	//	+------+--------+---------+----------------+
+	//
+	// Cases:
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// | Case | From  |   To   | Port |    Query    | Outcome  |               Reason                |
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// |    1 | EPID1 | DstID1 |   53 | cilium.io   | Allowed  |                                     |
+	// |    2 | EPID1 | DstID1 |   54 | cilium.io   | Rejected | Port 54 only allows example.com     |
+	// |    3 | EPID1 | DstID2 |   53 | cilium.io   | Rejected | DstID2 is not allowed in any policy |
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy_test.go]: I don't understand, the table on line 437 states that DstID2 is in fact allowed on port 53 to `cilium.io`, but this table says that such traffic is rejected.

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 11]
```diff
@@ -163,131 +206,325 @@ func (s *DNSProxyTestSuite) TearDownSuite(c *C) {
...
+				cachedFooSelector: api.L7Rules{
+					DNS: []api.PortRuleDNS{
+						{MatchName: "aws.amazon.com."},
+					},
+				},
+			},
+			{
+				// Rule
+				//	| All  | DstID2 |      53 | cilium.io      |
+				cachedFooSelector: api.L7Rules{
```
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy_test.go]: The comment above (and the other comments) use `DstID2` to indicate the L3 peer here, but it looks like it's using the same `cachedFooSelector` as for `DstID1`.

**Review by joestringer — CHANGES_REQUESTED** [review]
Nice work. There's just a couple of inconsistencies in the unit testing that are confusing, could you take a fresh look at those?

I left some other minor comments as well but I don't think any of them are consequential. Assuming you fix up the unit tests (or describe what I missed), LGTM.

**Comment by joestringer** [comment]
CI failure seems legitimate:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/15866/testReport/junit/Suite-k8s-1/11/K8sPolicyTest_Basic_Test_Redirects_traffic_to_proxy_when_no_policy_is_applied_with_proxy_visibility_annotation_Tests_DNS_proxy_visibility_without_policy/

File: pkg/fqdn/dnsproxy/proxy.go (hunk 8 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go:169]: I had started by keeping the L7DataMap around as well but I realised I don't ever use the DNS rules again, but only the regexes. I'm guessing it's safe to keep the CacheSelector reference since we do it in the redirect object anyway.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 7 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go:406]: There was but I can't find it. I actually had the same thought while redoing the tests. I've switched it to uint16, which is the more natural size anyway. 👍

Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go:406]: Oh, it's because Redirect.EndpointID is a uint64. I'm going to leave the upcast since it is safer if we ever change the endpoint ID size.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 9 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go:396]: good catch! Processing errors are reported one way whereas blocks are treated another (namely, the sendRefused processing time and error are included in NotifyOnDNSMsg). The .End(true/false) also needs to be consistent.
I've fixed things up in this file (I hope)

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 11 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy_test.go]: Oops, you are correct. I'll rename cacheFooSelector to Dst1Selector or something to make it clearer. I'll also add the Dst2 equivalents so this rule works (and fix the test cases)

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 10 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy_test.go]: Yup, this was a mistake. I'll fixup Dst2 per your other comment but this test should be for aws.amazon.com. It was intended to check for aliasing between L3 selectors (i.e. aws.amazon.com is only allowed for Dst1). The other tests are also a little off, since cilium.io is only allowed with Dst2... but Case 1 thinks it's allowed :/

Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy_test.go]: (the same for EPID2. The policies should say From: EPID1 in all cases)

File: pkg/fqdn/dnsproxy/proxy.go (hunk 7 above)
Comment (joestringer) [reply pkg/fqdn/dnsproxy/proxy.go:406]: We could probably do an `EndpointID` type to address this but this is good enough for now.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 8 above)
Comment (joestringer) [reply pkg/fqdn/dnsproxy/proxy.go:169]: Yeah, I think so. There was actually a bug in this area recently but Jarno's got a fix out for it: #9617

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go [hunk 12]
```diff
@@ -223,56 +300,54 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
...
-		p.allowed.Remove(reStr, endpointID)
+
+	epAllow, exists := p.allowed.getPortRulesForID(endpointID, destPort)
+	if !exists {
+		return false, nil
 	}
-	for _, reStr := range reStrToAdd {
-		p.allowed.Add(reStr, endpointID)
+
+	// FIXME: We see cachedSelectors with nil internals, possibly from visibility
```
Comment (raybejjani) [inline pkg/fqdn/dnsproxy/proxy.go]: @joestringer I'm a bit unsure about this. This is my best guess, but it feels like a hack. Is there a way to check here that it's a visibility policy? (instead of allowing blindly on a panic)

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go (hunk 12 above)
Comment (joestringer) [reply pkg/fqdn/dnsproxy/proxy.go]: Here's your problem:
https://github.com/cilium/cilium/blob/904b274e2c76edf232758126cce4b369bf356a3a/pkg/policy/visibility.go#L148

This looks like it should be `api.WildcardEndpointSelector` instead.

**Review by joestringer — CHANGES_REQUESTED** [review]
The defer is clearly masking over an input problem, and I'm hoping that my pointer in the related comment here will resolve it.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 12 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: Hmm, it wants a type that implements CachedSelector (this is missing GetSelections it seems). I'll try to figure out how to build the right objects. I'm not 100% clear on the path from one to the other. Thanks for finding it, though!

Comment (joestringer) [reply pkg/fqdn/dnsproxy/proxy.go]: Ah, right you are. Maybe we should consider creating a singleton `CachedWildcardEndpointSelector` which caches the selections of the `api.WildcardEndpointSelector` for use here.

Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: Yup, that's exactly what I tried. I wasn't sure what to pass into the constructor, though. It wants a "user". I'll look through the way we make others and figure it out. I agree that it's the righter thing to do (over the panic/recover, certainly).

Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: @joestringer Ok, I couldn't do it quite right but it self-consistent. The selector realises that it's a wildcard, and so avoids dereferencing itself (well, parts). It's a tad brittle because it relies on the current internal logic but I doubt that part will change (short-circuiting on wildcard).
I couldn't make this a global side-by-side with `api.WildcardEndpointSelector` because it relies on the policy-internal `wildcardSelectorKey`. I can move those about more if you think it's worth it. I can also make this part a separate PR

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go (hunk 12 above)
Comment (joestringer) [reply pkg/fqdn/dnsproxy/proxy.go]: The `user` part is the mechanism we would use when a new identity shows up in the cluster which matches the 'policy' so that we can push an incremental update down to the datapath.

The nice thing with the visibility policy is that it is supposed to match all traffic, so I think even if the exact plumbing is subtly different, this is likely to correctly match traffic. Looking at what you've done, it looks sufficient for now. We can take a note to explore more subtle cases like identity update to ensure that works as expected, as a follow-up to this PR.

**Review by joestringer — APPROVED** [review]

**Comment by raybejjani** [comment]
Oops, I didn't meant to ask for another review, joestringer. Ignore me!

File: pkg/fqdn/dnsproxy/proxy.go [hunk 13]
```diff
@@ -223,56 +300,45 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
...
-		reStrToAdd[i] = prepareNameMatch(reStrToAdd[i])
-	}
-	for i := range reStrToRemove {
-		reStrToRemove[i] = prepareNameMatch(reStrToRemove[i])
-	}
-
+// CheckAllowed checks endpointID, destPort, destID, and name against the rules
+// added to the proxy, and only returns true if this all match something that
+// was added (via SetAllowed) previously.
+func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID identity.NumericIdentity, name string) (allowed bool, err error) {
```
Comment (jrajahalme) [inline pkg/fqdn/dnsproxy/proxy.go:315]: I don't see any error returns, so maybe remove the second return value?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 14]
```diff
@@ -95,31 +106,95 @@ type DNSProxy struct {
...
-	allowed *regexpmap.RegexpMap
+	allowed perEPAllow
 
 	// rejectReply is the OPCode send from the DNS-proxy to the endpoint if the
 	// DNS request is invalid
 	rejectReply int
 }
 
+// perEPAllow maps EndpointIDs to ports + selectors + rules
+type perEPAllow map[uint64]portToSelectorAllow
```
Comment (jrajahalme) [inline pkg/fqdn/dnsproxy/proxy.go:127]: Maybe consider tracking ports, selectors, and rules by a SecurityID instead of by EndpointID? Seems we could share the state between endpoints that have the same SecurityID here. There would be some additional complexity though due to policies being created per EP. `policy.L4Filter.L7RulesPerEp` might be a proper place to keep this information. Note that there is an another (wip) PR (https://github.com/cilium/cilium/pull/9486) that generalizes `L7RulesPerEp` a little, allowing for computed state being stored there, not just the `api.L7Rules`.

**Review by jrajahalme — APPROVED** [review]
One suggestion for a follow-up: Maybe compute selector-based rules when the policy is first injected, so that the rules would be natively shared by all the Endpoints applying the selector policy in question?

The other comment is just a nit about unused return error value, removal of which would also remove some dead code for checking the error return.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 13 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go:315]: Good catch. This was a carry-over from an intermediate implementation. I'll remove it.

**Comment by raybejjani** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
